lib/java/com/google/android/material/lists/MaterialListBinder.java

```diff
@@ -0,0 +1,8 @@
+package com.google.android.material.lists;
+
+import androidx.recyclerview.widget.RecyclerView.ViewHolder;
+
+/** Binder similar to onBindViewHolder() but in interface callback format */
+public abstract class MaterialListBinder<T extends ViewHolder> {
+  public abstract void onBind(T viewHolder, int position);
+}
```

Comment:
Any reason for not including the type of data the `MaterialListBinder` binds to? Like this:

```java
public abstract class MaterialListBinder<T, VH extends RecyclerView.ViewHolder> {
  public abstract void onBind(@NonNull VH viewHolder, @NonNull T item);
}

Otherwise LGTM
```

Reply:
Glad you asked @bohsen ! The reason is because there is no standard data structure for lists (could be a list, linkedklist, array, map) and we want to let the user customize how they send data to the adapter. Even if we allowed iterable data structures the user would have little control.

Reply:
I think those are two different questions. @bohsen's is `Can we add the type of the item to the bind method in this interface?`. Yours is `Can we make an interface for MaterialListAdapter that doesn't restrict how the user sends data to the adapter?`. Right?

In theory we could make an interface that has a method to return a size, and another to return an object of a particular type given the index. We'd be able to use that in the current implementation rather than passing the size directly. This would allow us to add a type, but keep the current flexibility.

I actually think an interable is the most common case.. And it's pretty easy to convert some other data type to be interable to pass in, so I'm not sure how much of an advantage there is to keep it more open.

Reply:
@ethanhsuhsu @cketcham I think the API should be picky about what to accept as input in regards to data type to withstand changes later on. And I believe the `List` interface would be the right choice, as this is the case in almost all cases. Plus other `Collection` types are easily transformed into `List`s with `asList/toList` extensions/static utility methods. 

> In theory we could make an interface that has a method to return a size, and another to return an object of a particular type given the index. We'd be able to use that in the current implementation rather than passing the size directly. This would allow us to add a type, but keep the current flexibility.

If you look at how `androidx.recyclerview.widget.ListAdapter` is implemented, this is how they've done (using a class though - `AsyncListDiffer`). This also brings the benefit of being able to "install" a changelistener. It actually might be worth considering if `MaterialListAdapter` should reuse from `ListAdapter`/`AsyncListDiffer` if this ends up having a dependency on `androidx.recyclerview` anyway, as mentioned [here](https://github.com/material-components/material-components-android/issues/81#issuecomment-556043356). Also `androidx.recyclerview` version 1.1.0 was just released as stable.

> could be a `List`, `LinkedList`, `Array`, **`Map`**

IMO support for `Map`s is unnecessary and I also don't believe this would support it that well.  The main argument being that the `position` integer parameter would have to map to the value. In my experience integers are rarely used as keys in `Map`s. Please excuse me if I'm missing something in that regard.

But all of this is probably targeted at dynamic data support, which @ethanhsuhsu already mentioned [here](https://github.com/material-components/material-components-android/pull/798#discussion_r352803979) isn't the purpose of this PR. So maybe merge this  PR and continue with the next step in the implementation (Update the list item xml layouts to be more consistent with Material standards), if all other comments have been addressed? And then people interested in this can ponder on this until the next PR.

Btw. Sorry for not following best practice earlier in regard to my review. That was sloppy on my behalf.

Reply:
I also agree `List` is probably the most common, and it's pretty easy to convert to and from a list. I'd be ok with either `List` or `Iterable`.

I understand the purpose of this PR isn't targeted at dynamic data support.. but I don't know that the current api (passing only the size to the adapter) makes the most sense considering most often we'll want to support a `List` with objects with a consistent type. It might be better to take in a list in the constructor.. Or use `androidx.recyclerview.widget.ListAdapter` like you suggest to have the option of doing diffs easily.

Reply:
Lists, arrays and custom types are now supported!